securebanking-openbanking-uk-rs-simulator/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/obie/payment/v3_0/domesticpayments/DomesticPaymentsApiController.java

```diff
@@ -0,0 +1,130 @@
+import static com.forgerock.securebanking.openbanking.uk.rs.converter.payment.FRWriteDomesticConverter.toFRWriteDomestic;
+import static com.forgerock.securebanking.openbanking.uk.rs.persistence.document.payment.FRSubmissionStatus.PENDING;
+
+@Controller("DomesticPaymentsApiV3.0")
+@Slf4j
+public class DomesticPaymentsApiController implements DomesticPaymentsApi {
+
+    private final DomesticPaymentSubmissionRepository domesticPaymentSubmissionRepository;
+    private final PaymentSubmissionValidator paymentSubmissionValidator;
+
+    public DomesticPaymentsApiController(DomesticPaymentSubmissionRepository domesticPaymentSubmissionRepository,
+                                         PaymentSubmissionValidator paymentSubmissionValidator) {
+        this.domesticPaymentSubmissionRepository = domesticPaymentSubmissionRepository;
+        this.paymentSubmissionValidator = paymentSubmissionValidator;
+    }
+
+    @Override
+    public ResponseEntity<OBWriteDomesticResponse1> createDomesticPayments(@Valid OBWriteDomestic1 obWriteDomestic1,
+                                                                           String xFapiFinancialId,
+                                                                           String authorization,
+                                                                           String xIdempotencyKey,
+                                                                           String xJwsSignature,
+                                                                           DateTime xFapiCustomerLastLoggedTime,
+                                                                           String xFapiCustomerIpAddress,
+                                                                           String xFapiInteractionId,
+                                                                           String xCustomerUserAgent,
+                                                                           HttpServletRequest request,
+                                                                           Principal principal
+    ) throws OBErrorResponseException {
+        log.debug("Received payment submission: '{}'", obWriteDomestic1);
+
+        // TODO - before we get this far, the IG will need to:
+        //      - verify the consent status
+        //      - verify the payment details match those in the payment consent
+        //      - verify security concerns (e.g. detached JWS, access token, roles, MTLS etc.)
+
+        paymentSubmissionValidator.validateIdempotencyKeyAndRisk(xIdempotencyKey, obWriteDomestic1.getRisk());
+
+        FRWriteDomestic frDomesticPayment = toFRWriteDomestic(obWriteDomestic1);
+        log.trace("Converted to: '{}'", frDomesticPayment);
+        FRDomesticPaymentSubmission frPaymentSubmission = FRDomesticPaymentSubmission.builder()
+                // TODO - openbanking-aspsp uses the consent id - is this for convenience? Could it be a problem?
```

Comment:
Any thoughts why this might be?

Reply:
The openbanking spec define the `consentId` field and it being mapped like as the object Id stored in mongo,  indeed it's a mandatory property:
```java
@NotNull
    @Size(
        min = 1,
        max = 128
    )
    @ApiModelProperty(
        required = true,
        value = "OB: Unique identification as assigned by the ASPSP to uniquely identify the consent resource."
    )
    public String getConsentId() {
        return this.consentId;
    }
``` 
I'm not sure how IDM will store this object but probably has the id for the stored object, otherwise we can create on the fly a unique value to store it in the id field, and have the expected id in the consent flow response.

In short isn't a convenience it's mandatory and can't be a problem, unless the response haven't that `consentId` field.

Reply:
Indeed the `consentId` is mandatory within `OBWriteDataDomestic1` (that's within the provided `OBWriteDomestic1`) and it's stored within `FRWriteDataDomestic` (within `FRWriteDomestic`) in Mongo which is all fine, but that's not what I'm talking about here.

The consent will have to exist in IDM and IG will need to ensure it has been authorised before sending the payment submission here.

The thing that's odd is that in `openbanking-aspsp`, the primary key of the submission is set to the `consentId`, rather than having its own generated value. The submission already contains the consentId (as described above), but I'm talking about the ID of the submission here. Normally, it would have its own unique ID.

Reply:
I think that should have its own unique ID to avoid ID conflicts.

Reply:
That's what I thought, but I wonder why it doesn't in openbanking-aspsp 🤔

Reply:
I think this may be the answer. When the call to `idempotentSave()` is made, it checks that a paymentSubmission doesn't already exist:
```
public T idempotentSave(T paymentSubmission) throws OBErrorResponseException {
   Optional<T> isPaymentSubmission = repo.findById(paymentSubmission.getId());
   if (isPaymentSubmission.isPresent()) {
      log.info("A payment with this payment id '{}' was already found. Checking idempotency key.", isPaymentSubmission.get().getId());
      IdempotencyValidator.validateIdempotencyRequest(paymentSubmission, isPaymentSubmission.get());
      log.info("Idempotent request is valid. Returning [201 CREATED] but take no further action.");
      return isPaymentSubmission.get();
```
I guess this is a way to ensure there's only one `paymentSubmission` per consent, but the same could have been achieved by first checking if a submission exists for the `consentId` in question - i.e:
```
Optional<T> isPaymentSubmission = repo.findByConsentId(paymentSubmission.getConsentId());
```
I'll have to add `getConsentId()` to the `PaymentSubmission` interface. Each submission object would know how to get its own the consentId internally.

Normally, I'd always advocate generating an ID for a new object each time, but I'm slightly caught in two minds here. On balance I still think it's better to create a new ID to avoid any confusion between a `consentId` and a `submissionId`.